File: tree/main/samples/agent/adk/gemini_enterprise/agent_engine/agent_sdks/kotlin/src/main/kotlin/com/google/a2ui/core/parser/PayloadFixer.kt
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.a2ui.core.parser

import java.util.logging.Logger
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

/** Validates and applies autofixes to raw JSON strings. */
object PayloadFixer {

  private val logger = Logger.getLogger(PayloadFixer::class.java.name)

  /**
   * Parses and applies autofixes to a raw JSON string and returns the parsed payload as a list of
   * components.
   *
   * @param payload The raw JSON string from the LLM.
   * @return A parsed and potentially fixed payload (array of JsonElements).
   */
  fun parseAndFix(payload: String): List<JsonElement> {
    val normalizedPayload = normalizeSmartQuotes(payload)
    return try {
      parse(normalizedPayload)
    } catch (e: Exception) {
      logger.warning("Initial A2UI payload validation failed: ${e.message}")
      parse(removeTrailingCommas(normalizedPayload))
    }
  }

  /**
   * Replaces smart (curly) quotes with standard straight quotes.
   *
   * @param jsonStr The raw JSON string from the LLM.
   * @return A string with curly quotes replaced by straight quotes.
   */
  fun normalizeSmartQuotes(jsonStr: String): String {
    return jsonStr
      .replace('\u201C', '"') // “
      .replace('\u201D', '"') // ”
      .replace('\u2018', '\'') // ‘
      .replace('\u2019', '\'') // ’
  }

  private fun parse(payload: String): List<JsonElement> =
    try {
      when (val element = Json.parseToJsonElement(payload)) {
        is JsonArray -> element.toList()
        is JsonObject -> {
          logger.info("Received a single JSON object, wrapping in a list for validation.")
          listOf(element)
        }
        else ->
          throw IllegalArgumentException(
            "Payload must be a JSON Array or Object, got: ${element::class.simpleName}"
          )
      }
    } catch (e: Exception) {
      logger.severe("Failed to parse JSON: ${e.message}")
      throw IllegalArgumentException("Failed to parse JSON: ${e.message}", e)
    }

  /**
   * Attempts to remove trailing commas from a JSON string.
   *
   * This implementation is quote-aware and will not modify commas inside strings.
   *
   * @param jsonStr The raw JSON string from the LLM.
   * @return A potentially fixed JSON string.
   */
  fun removeTrailingCommas(jsonStr: String): String {
    val result = StringBuilder()
    var inString = false
    var i = 0
    var lastCommaIndex = -1

    while (i < jsonStr.length) {
      val c = jsonStr[i]

      if (c == '"' && (i == 0 || jsonStr[i - 1] != '\\')) {
        inString = !inString
      }

      if (!inString) {
        when {
          c == ',' -> lastCommaIndex = result.length
          (c == ']' || c == '}') && lastCommaIndex != -1 -> {
            val contentBetween = result.substring(lastCommaIndex + 1)
            if (contentBetween.isBlank()) {
              result.deleteCharAt(lastCommaIndex)
            }
            lastCommaIndex = -1
          }
          !c.isWhitespace() -> lastCommaIndex = -1
        }
      }

      result.append(c)
      i++
    }

    val fixedJson = result.toString()
    if (fixedJson != jsonStr) {
      logger.warning("Detected trailing commas in LLM output; applied robust autofix.")
    }
    return fixedJson
  }
}

File: agent_sdks/cpp/src/parser/payload_fixer.cc
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "a2ui/parser/payload_fixer.h"
#include <regex>
#include <iostream>
#include <fstream>

namespace a2ui {

std::string normalize_smart_quotes(const std::string& json_str) {
    std::string s = json_str;
    // Replace smart quotes with standard quotes (UTF-8 hex codes)
    s = std::regex_replace(s, std::regex("\xE2\x80\x9C"), "\"");
    s = std::regex_replace(s, std::regex("\xE2\x80\x9D"), "\"");
    s = std::regex_replace(s, std::regex("\xE2\x80\x98"), "'");
    s = std::regex_replace(s, std::regex("\xE2\x80\x99"), "'");
    
    return s;
}

std::string remove_trailing_commas(const std::string& json_str) {
    // Fix trailing commas: identifying commas followed by optional whitespace and a closing bracket (]) or brace (}).
    std::regex pattern(R"(,(?=\s*[\]}]))");
    std::string fixed = std::regex_replace(json_str, pattern, "");

    return fixed;
}

static nlohmann::json _parse(const std::string& payload) {
    try {
        nlohmann::json a2ui_json = nlohmann::json::parse(payload);
        if (!a2ui_json.is_array()) {
            a2ui_json = nlohmann::json::array({a2ui_json});
        }
        return a2ui_json;
    } catch (const std::exception& e) {
        throw std::runtime_error("Failed to parse JSON: " + std::string(e.what()));
    }
}

nlohmann::json parse_and_fix(const std::string& payload) {
    std::string normalized_payload = normalize_smart_quotes(payload);
    try {
        return _parse(normalized_payload);
    } catch (const std::exception& e) {
        std::string updated_payload = remove_trailing_commas(normalized_payload);
        return _parse(updated_payload);
    }
}

} // namespace a2ui
